### PR TITLE
[SPARK-23167][SQL] Add TPCDS queries v2.7 in TPCDSQuerySuite

### PR DESCRIPTION
**Pull request by maropu**
## What changes were proposed in this pull request?
This pr added TPCDS v2.7 (latest) queries in `TPCDSQuerySuite` because the current `TPCDSQuerySuite` tests older one (v1.4) and some queries are different from v1.4 and v2.7. Since the original v2.7 queries have the syntaxes that Spark cannot parse, I changed these queries in a following way:

 - [date] + 14 days -> date + `INTERVAL` 14 days
 - [column name] as "30 days" -> [column name] as \`30 days\`
 - Fix some syntax errors, e.g., missing brackets

## How was this patch tested?
Added tests in `TPCDSQuerySuite`.

### CODE REVIEW

File: sql/core/src/test/scala/org/apache/spark/sql/TPCDSQuerySuite.scala [hunk 1]
```diff
@@ -339,6 +340,30 @@ class TPCDSQuerySuite extends BenchmarkQueryTest {
...
+    "q28", "q29", "q30", "q31", "q32", "q33", "q34", "q35", "q35a", "q36", "q36a", "q37", "q38",
+    "q39_1", "q39_2", "q40", "q41", "q42", "q43", "q44", "q45", "q46", "q47", "q48", "q49",
+    "q50", "q51", "q51a", "q52", "q53", "q54", "q55", "q56", "q57", "q58", "q59",
+    "q60", "q61", "q62", "q63", "q64", "q65", "q66", "q67", "q67a", "q68", "q69",
+    "q70", "q70a", "q71", "q72", "q73", "q74", "q75", "q76", "q77", "q77a", "q78", "q79",
+    "q80", "q80a", "q81", "q82", "q83", "q84", "q85", "q86", "q86a", "q87", "q88", "q89",
+    "q90", "q91", "q92", "q93", "q94", "q95", "q96", "q97", "q98", "q99")
+
+  tpcdsQueriesV2_7_0.foreach { name =>
+    val queryString = resourceToString(s"tpcds-v2.7.0/$name.sql",
```
Comment (gatorsmile) [inline sql/core/src/test/scala/org/apache/spark/sql/TPCDSQuerySuite.scala]: Could we just add the queries that are different from the v1.4 version?

Comment (maropu) [reply sql/core/src/test/scala/org/apache/spark/sql/TPCDSQuerySuite.scala]: ok

Comment (dongjoon-hyun) [reply sql/core/src/test/scala/org/apache/spark/sql/TPCDSQuerySuite.scala]: @maropu . It's great to have v2.7.
Could you check the schema too?
For example, we had better update [the following in the schema](https://github.com/apache/spark/pull/20343/files#diff-38fa80d1dc9860f07e135dd02d259269R247)?
```
-        |`web_country` STRING, `web_gmt_offset` STRING, `web_tax_percentage` DECIMAL(5,2))
+        |`web_country` STRING, `web_gmt_offset` DECIMAL(5,2), `web_tax_percentage` DECIMAL(5,2))
```

Comment (maropu) [reply sql/core/src/test/scala/org/apache/spark/sql/TPCDSQuerySuite.scala]: ok, thanks.

Comment (gatorsmile) [reply sql/core/src/test/scala/org/apache/spark/sql/TPCDSQuerySuite.scala]: This is only related to test cases. Thus, it is fine Spark 2.3 release does not have it. You can do it in this PR.

Actually, this PR can be merged as long as we can fix all the issues.

Comment (maropu) [reply sql/core/src/test/scala/org/apache/spark/sql/TPCDSQuerySuite.scala]: ok

**Comment by maropu** [comment]
retest this please.

**Comment by maropu** [comment]
We need to update `TPCDSQueryBenchmark`, too? I think we could replace the update queries there.

File: sql/core/src/test/resources/tpcds-v2.7.0/q11.sql [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+         ,c_birth_country
+         ,c_login
+         ,c_email_address
+         ,d_year
+         )
+  select  
+                  t_s_secyear.customer_id
+                 ,t_s_secyear.customer_first_name
+                 ,t_s_secyear.customer_last_name
+                 ,t_s_secyear.customer_email_address
```
[the remark this answers is not in the document]
Comment (gatorsmile) [reply sql/core/src/test/resources/tpcds-v2.7.0/q11.sql]: If possible, could you also use the original SQL query files with the same styles/indents?

File: sql/core/src/test/scala/org/apache/spark/sql/TPCDSQuerySuite.scala [hunk 3]
```diff
@@ -244,7 +244,7 @@ class TPCDSQuerySuite extends BenchmarkQueryTest {
         |`web_company_id` INT, `web_company_name` STRING, `web_street_number` STRING,
         |`web_street_name` STRING, `web_street_type` STRING, `web_suite_number` STRING,
         |`web_city` STRING, `web_county` STRING, `web_state` STRING, `web_zip` STRING,
-        |`web_country` STRING, `web_gmt_offset` STRING, `web_tax_percentage` DECIMAL(5,2))
+        |`web_country` STRING, `web_gmt_offset` DECIMAL(5,2), `web_tax_percentage` DECIMAL(5,2))
```
[the remark this answers is not in the document]
Comment (dongjoon-hyun) [reply sql/core/src/test/scala/org/apache/spark/sql/TPCDSQuerySuite.scala]: Thank you a lot, @maropu and @gatorsmile !

**Comment by maropu** [comment]
ok, I'll try and check. Just s sec.

File: sql/core/src/test/scala/org/apache/spark/sql/TPCDSQuerySuite.scala [hunk 4]
```diff
@@ -339,6 +340,24 @@ class TPCDSQuerySuite extends BenchmarkQueryTest {
     }
   }
 
+  val tpcdsQueriesV2_7_0 = Seq(
```
[the remark this answers is not in the document]
Comment (maropu) [reply sql/core/src/test/scala/org/apache/spark/sql/TPCDSQuerySuite.scala]: ok

File: sql/core/src/test/resources/tpcds-v2.7.0/q11.sql (hunk 2 above)
[the remark this answers is not in the document]
Comment (maropu) [reply sql/core/src/test/resources/tpcds-v2.7.0/q11.sql]: ok, I'll do as much as possible. one question; I think the sql format is not consistent in all the files, e.g., `SQLQueryTestSuite`. Probably, I think uppercase letters for SQL reserved words and 2 indents seem to be de-facto, but we don't have any format rule for that anywhere, right? we'd better to write the rule somewhere? We don't need to re-format existing code though, IMHO we'd better to make them consistent in future prs.

[the remark this answers is not in the document]
Comment (gatorsmile) [reply sql/core/src/test/resources/tpcds-v2.7.0/q11.sql]: Regarding a keywords capitalization rule, this is just for readability. We do not enforce it, but it is preferred.

**Comment by maropu** [comment]
I opened a new pr to support `[date] + 14 days`: https://github.com/apache/spark/pull/20433

**Comment by maropu** [comment]
I checked all the queries again and I found that some queries (q6, q11, q20, q22, q24, q34, q35, q47, q49, q57, q64, q72, q74, q75, q78, q98) only have minor changes (See the comments to point out the  changes). So, how about directly applying these changes in `sql/core/src/test/resources/tpcds`?

**Comment by maropu** [comment]
retest this please.

**Review by dongjoon-hyun — APPROVED** [review]
+1, LGTM.

**Comment by gatorsmile** [comment]
@maropu Yeah. As long as the queries are different, we should keep both versions. This is to help the others understand we fully support TPC-DS queries without the changes. Thanks!

**Comment by gatorsmile** [comment]
Thanks for submitting the PR https://github.com/apache/spark/pull/20433. It sounds like there are still some test failure. Will review it after 2.3 release. Thanks!

**Comment by maropu** [comment]
ok, I'll fix soon. Many Thanks!

**Comment by kiszk** [comment]
ping @gatorsmile

**Comment by maropu** [comment]
We need to review #20433 first

**Comment by maropu** [comment]
A support of the optional intervals will be planed in 3.x (#20433), so is it okay to restart this again? @gatorsmile

**Comment by gatorsmile** [comment]
retest this please

**Comment by gatorsmile** [comment]
Thanks! Merged to master

**Pull request closed without merging**